doc/design-sparse.md

```diff
@@ -0,0 +1,52 @@
+   - Coordinates for event data would need to be named, since they are specific for their data --- data for, e.g., sample and vanadium runs would *not* share these coordinates.
+     This furthermore implies that we must allow for mismatches in the sparse dimension, i.e., a single dataset can hold two variables with sparse dimension that have mismatching sparse extents.
+   - Does this imply that we cannot easily store event- and histogram-data in the same dataset (with the same name)?
+     Can we have some variables where, e.g., `Dim.Tof` is dense, whereas there are others where it is sparse? The simplest example would be storing and (unnamed) `Coord.Tof` in a dataset together with events, which could later be binned to that coordinate, similar to Mantid's `EventWorkspace`.
+     - Could support sparse dimension mismatching dense dimension extent.
+
+4. Access.
+   - In general, `Variable::span<T>` can (and usually should) not work for sparse data.
+     We do not want to expose the sparse storage layout on the interface, since this would prevent future changes.
+     In particular, we should avoid being forced to store sparse data as a single contiguous block.
+   - Use something like `Variable::sparseProxy<T>`, returning `SparseProxy<T>`, which has `SparseProxy<T>::span(const scipp::index i)`, with `i` in range `[0,...,volume_of_non_sparse_dimensions]`. The latter returns a span to the range of items of type `T` (the 1D list of values for a particular event list). That is, we do commit to a contiguous layout of data within a single "event list".
+
+5. Slicing.
+   - We cannot slice in the sparse dimension.
+   - Investigate whether a fully sliced variable can appear as non-sparse, i.e., when all but the sparse dimension have been removed.
+   - Special value-based slicing methods could be supported, but would obviously be less efficient than slicing dense dimensions.
+
+6. Operations.
+   - `sort`, `concatenate`, `filter`, ... work well and are useful for sparse dimensions.
+   - For operations between two sparse variables with matching dimensions we can apply binary operations as usual.
+     Note that is is different from Mantid's behavior of translating binary operations into special other operations such as concatenation.
+   - For operations between two sparse variables with mismatching dimensions, we need to provide alternatives.
+     - use `concatenate` instead of `+`.
+     - instead of `-`, add weights (negative for second operand) and then concatenate.
+     - ...
+   - Operation between dense and sparse data, assuming the dense data does not depend on the sparse dimensions work naturally.
+     The result is sparse data, i.e, the dense data is simply broadcast into the sparse dimensions.
+     This makes implementation of, e.g., unit conversion convenient.
+
+7. Data layout. With the access described in 4.), we can choose to store sparse data in various ways.
```

Comment:
If I understand the design proposal - the sparse dimensions could get very large. For large number of events in a small region (quite typical) is it worth considering some kind of tolerance with the sparse dimensions?

Reply:
Can you elaborate?

Reply:
I'm thinking for events corresponding to some strong bragg peak or other scattering feature. This would be particularly acute if summing (concatenated) over several detectors or a monitor. The dimension TOF precision may be more than you need, in which case you may be recording many independent TOF sparse coordinates that could otherwise be treated the same.

Reply:
So you are referring to something like compressed events in Mantid (merging values from nearby positions along the sparse dimension)? You can implement this here as well, but I do not think that it should be built into the data type. Compression should just be applied manually.

Reply:
Yes, other than recording what compression might have been applied. I guess that's more like a question for the history functionality.